### ISSUE TITLE
What are the requirements for automatic instrumentation?

### ISSUE THREAD

**Issue by mrtnzlml** [DataDog/dd-trace-js#560]
I was trying to make `dd-trace` work according to documentation but seems like it only works for Node.js native modules. In our case, it doesn't work for NPM modules like GraphQL. So I started to dig into the source code and seems like this library requires not only `package.json` but also `node_modules` folder. And we don't have either of this in production. So my questions are:

1) What are the requirements for automatic instrumentation? Seems like it's not documented.
2) How to do manual instrumentation? For example, I know we use GraphQL so I'd like to force it without `package.json` and other magic stuff.

Thank you very much for your answers.

**Comment by rochdev** [comment]
The only reason automatic instrumentation depends on `node_modules` is to match the path in the instrumented module folder. We can probably adjust that for other folders. Where do you store the modules if not in `node_modules`?

**Comment by mrtnzlml** [comment]
We are deploying from monorepo and it's not so simple to just take the repo, transpile it and deploy it. So we bundle all our Node.js code into a single file and just execute it. It simplifies many things and we are quite happy with this setup. But it means there are not `package.json` or `node_modules.

It would be great if I had the ability to manually instrument some modules. I understand that automatic instrumentation is great for quick starts but I'd like to have this more under control (especially in this situation when the automatic way doesn't work).

Thanks for any advice on how to make it work.

**Comment by rochdev** [comment]
In that case you will indeed have to rely on manual instrumentation. You can find documentation [here](https://datadog.github.io/dd-trace-js/#manual-instrumentation). There is also a new API that is simpler to use but not yet documented that you can find [here](https://datadog.github.io/dd-trace-js/interfaces/tracer.html#trace).

What do you use to bundle the code together? Do you think it would make sense to support bundlers to get automatic instrumentation for use case like yours? We're also thinking of exposing module specific APIs eventually to make it easier to do manual instrumentation on supported modules when automatic instrumentation is not possible or wanted.

**Comment by mrtnzlml** [comment]
Thanks! It would be really great to be able to reuse some parts of GraphQL tracer like function `wrap{Execute|Parse|Validate}` for example.

> Do you think it would make sense to support bundlers to get automatic instrumentation for use case like yours?

Honestly, I don't think so. We currently use Webpack because it makes our life so much easier. But we could eventually change it to something else, why not. It's a bit weird to use Webpack on a server, I know. :) Automatic instrumentation is good for most of the projects but I wouldn't try to catch every use-case. It's better to be able to easily opt-out from it.

**Comment by rochdev** [comment]
Ok, so for now I would say to use the manual instrumentation and feel free to update this thread with any question or pain point. This will also help us to design these module specific APIs.

**Comment by mrtnzlml** [comment]
> There is also a new API that is simpler to use but not yet documented that you can find here.

@rochdev Do you maybe have some example of how to use it correctly?

**Comment by rochdev** [comment]
There are a few examples in https://github.com/DataDog/dd-trace-js/pull/457, mostly showing the equivalent code using the low level API.

**Comment by BrunoBerisso** [comment]
I'm facing a similar problem with the automatic instrumentation.

I need to add tracing for a bunch of microservices using express but the codebase is big and use the automatic instrumentation it's really cumbersome.

I believe I could trace the requests by hand but the automatic process seems to do a lot more. Isn't there an API to manually trigger the same process for patching supported modules?

**Comment by rochdev** [comment]
@BrunoBerisso Could you describe exactly what the problem is in your case? Why is it not possible for you to instantiate the tracer before any other module?

**Comment by BrunoBerisso** [comment]
I'm working on a codebase with many shared modules, one of those is for instrumentation which already has StatsD integrated so the consumers already have an _require - init - use_ lifecycle for the module. This module is consumed by many other modules to build services (microservices) and isomorphic apps.

Because of the automatic instrumentation, I need to make changes on almost all the consumers of my module to make sure things load in the right order. If I had some API like the one in my PR the changes would be much simpler, or even unneeded.

There is also another detail about the load order, to get the settings to setup the tracer I need to use a separate module that gets everything from a server. This turns a sync load process to async making the lifecycle change more traumatic than just reorder lines.

Does it make sense? I'm answering your question?

**Comment by rochdev** [comment]
@BrunoBerisso Just to make sure I understand properly, do you mean that you have a flow something like the following?

1) Start an asynchronous request to get the configuration.
2) Continue execution which will result in all the `require` statements to be executed synchronously
3) Initialize the tracer with the configuration when the response from 1 is received, at which point all other `require` statements have already been run.

If that's the case, it should still be possible to call `tracer.use()` to patch the modules at that point, but there is no guarantee that this will be stable since there are cases where we need to hook into the internal modules before the main module is loaded.

**Comment by BrunoBerisso** [comment]
Yes, that's one of the worst cases. There are others when it's just annoying.

I try `tracer.use()` but it failed to patch Express. I get a _"fix"_ for that but I thought that a change on that method would have much fewer chances of getting through that add a new method that does the patching because of the backward compatibility issues it might have.

**Comment by rochdev** [comment]
I think there are 2 things to be addressed here:

1) If the new method works but not `tracer.use()` I would consider this a bug. `tracer.use()` should work even after initialization of the module, as long as it's done right after `tracer.init()` since the latter calls a `tracer.use()` equivalent internally for every plugins. The reason we document a specific flow is because as mentioned above trying to patch a module lazily can cause many issues.
2) There should be a plugin API that allow using the plugins manually without automatic instrumentation. This is definitely something we want to happen but there are a few changes needed in the project first.

If you can modify `tracer.use()` to support your use case, this would definitely be best as it means a better out of the box experience for anyone else who would otherwise face the same issue as you did. However, loading plugins lazily should be done with the understanding that this might now work with every module. If at all possible, initializing the tracer before everything else is definitely the recommended approach to avoid any issues.

I will try to think of a better way we could support your use case without the aforementioned issue. Maybe if we can add some kind of hook that would defer all `require` calls until the tracer is initialized then we can make sure that the correct order is kept while being minimally intrusive code-wise.
